File: web/src/api/endpoints/group-health.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '../client';
import { logger } from '@/lib/logger';
import { useGroupHealthEnabled } from './setting';
import type { GroupMode } from './group';

export type GroupHealthStatus = 'running' | 'success' | 'partial' | 'failed';
export type GroupHealthAttemptStatus = 'success' | 'failed' | 'skipped';
export type GroupHealthProbeMode = 'standard' | 'full';
export type GroupHealthMembershipState = 'active' | 'excluded' | 'missing';

export interface GroupHealthAttempt {
    id: number;
    snapshot_id: number;
    group_item_id: number;
    channel_id: number;
    channel_name: string;
    channel_key_id: number;
    key_remark: string;
    model_name: string;
    priority: number;
    weight: number;
    status: GroupHealthAttemptStatus;
    http_status: number;
    duration_ms: number;
    error_message: string;
    membership_state: GroupHealthMembershipState;
}

export interface GroupHealthExcludedItem {
    id: number; group_id: number; channel_id: number; channel_name: string; channel_enabled: boolean;
    model_name: string; priority: number; weight: number; excluded_at?: string | null;
    excluded_by_attempt_id?: number | null; http_status: number; duration_ms: number; error_message: string;
}

export interface GroupHealthRecoveryResult {
    item_id: number; restored: boolean; active_item_count: number;
    probe: { success: boolean; http_status: number; duration_ms: number; error_message: string };
}

export interface GroupHealthBatchRecoveryResult {
    total: number;
    restored_count: number;
    failed_count: number;
    active_item_count: number;
    results: GroupHealthRecoveryResult[];
}

export interface GroupHealthSnapshot {
    id: number;
    group_id: number;
    group_name: string;
    group_mode: GroupMode;
    probe_mode: GroupHealthProbeMode;
    request_model: string;
    status: GroupHealthStatus;
    started_at: string;
    finished_at?: string | null;
    duration_ms: number;
    successful_channel_id?: number | null;
    message: string;
    attempts: GroupHealthAttempt[];
}

export interface GroupHealthGroupView {
    group_id: number;
    group_name: string;
    group_mode: GroupMode;
    active_item_count: number;
    excluded_items: GroupHealthExcludedItem[];
    latest?: GroupHealthSnapshot | null;
}

export type RunGroupHealthAccepted = {
    group_id?: number;
    all_groups?: boolean;
    probe_mode?: GroupHealthProbeMode;
};

export type RunGroupHealthRequest = {
    groupId: number;
    probeMode?: GroupHealthProbeMode;
};

export type RunAllGroupHealthRequest = {
    probeMode?: GroupHealthProbeMode;
};

function normalizeAttempt(attempt: Partial<GroupHealthAttempt>): GroupHealthAttempt {
    return {
        id: typeof attempt.id === 'number' ? attempt.id : 0,
        snapshot_id: typeof attempt.snapshot_id === 'number' ? attempt.snapshot_id : 0,
        group_item_id: typeof attempt.group_item_id === 'number' ? attempt.group_item_id : 0,
        channel_id: typeof attempt.channel_id === 'number' ? attempt.channel_id : 0,
        channel_name: attempt.channel_name ?? '',
        channel_key_id: typeof attempt.channel_key_id === 'number' ? attempt.channel_key_id : 0,
        key_remark: attempt.key_remark ?? '',
        model_name: attempt.model_name ?? '',
        priority: typeof attempt.priority === 'number' ? attempt.priority : 0,
        weight: typeof attempt.weight === 'number' ? attempt.weight : 0,
        status: attempt.status === 'success' || attempt.status === 'failed' || attempt.status === 'skipped'
            ? attempt.status
            : 'failed',
        http_status: typeof attempt.http_status === 'number' ? attempt.http_status : 0,
        duration_ms: typeof attempt.duration_ms === 'number' ? attempt.duration_ms : 0,
        error_message: attempt.error_message ?? '',
        membership_state: attempt.membership_state === 'excluded' || attempt.membership_state === 'missing' ? attempt.membership_state : 'active',
    };
}

function normalizeSnapshot(snapshot: Partial<GroupHealthSnapshot> | null | undefined): GroupHealthSnapshot | null {
    if (!snapshot) return null;
    return {
        id: typeof snapshot.id === 'number' ? snapshot.id : 0,
        group_id: typeof snapshot.group_id === 'number' ? snapshot.group_id : 0,
        group_name: snapshot.group_name ?? '',
        group_mode: typeof snapshot.group_mode === 'number' ? snapshot.group_mode : 1,
        probe_mode: snapshot.probe_mode === 'full' ? 'full' : 'standard',
        request_model: snapshot.request_model ?? '',
        status: snapshot.status === 'running' || snapshot.status === 'success' || snapshot.status === 'partial' || snapshot.status === 'failed'
            ? snapshot.status
            : 'failed',
        started_at: snapshot.started_at ?? '',
        finished_at: snapshot.finished_at ?? null,
        duration_ms: typeof snapshot.duration_ms === 'number' ? snapshot.duration_ms : 0,
        successful_channel_id: typeof snapshot.successful_channel_id === 'number' ? snapshot.successful_channel_id : null,
        message: snapshot.message ?? '',
        attempts: (snapshot.attempts ?? []).map(normalizeAttempt),
    };
}

function normalizeView(view: Partial<GroupHealthGroupView>): GroupHealthGroupView {
    return {
        group_id: typeof view.group_id === 'number' ? view.group_id : 0,
        group_name: view.group_name ?? '',
        group_mode: typeof view.group_mode === 'number' ? view.group_mode : 1,
        active_item_count: typeof view.active_item_count === 'number' ? view.active_item_count : 0,
        excluded_items: (view.excluded_items ?? []).map((item) => ({
            id: item.id ?? 0, group_id: item.group_id ?? 0, channel_id: item.channel_id ?? 0,
            channel_name: item.channel_name ?? '', channel_enabled: item.channel_enabled ?? false,
            model_name: item.model_name ?? '', priority: item.priority ?? 0, weight: item.weight ?? 0,
            excluded_at: item.excluded_at ?? null, excluded_by_attempt_id: item.excluded_by_attempt_id ?? null,
            http_status: item.http_status ?? 0, duration_ms: item.duration_ms ?? 0, error_message: item.error_message ?? '',
        })),
        latest: normalizeSnapshot(view.latest),
    };
}

function invalidateGroupHealth(queryClient: ReturnType<typeof useQueryClient>) {
    queryClient.invalidateQueries({ queryKey: ['group-health', 'list'] });
    queryClient.invalidateQueries({ queryKey: ['groups', 'list'] });
    queryClient.invalidateQueries({ queryKey: ['group-health', 'detail'] });
}

export function useExcludeGroupHealthAttempt() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: ({ groupId, attemptId, allowEmpty = false }: { groupId: number; attemptId: number; allowEmpty?: boolean }) =>
            apiClient.post<GroupHealthGroupView>(`/api/v1/group/health/${groupId}/attempts/${attemptId}/exclude`, { allow_empty: allowEmpty }),
        onSuccess: () => invalidateGroupHealth(queryClient),
        onError: (error) => logger.error('group health exclusion failed:', error),
    });
}

export function useExcludeAllFailedGroupHealthAttempts() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: ({ groupId, allowEmpty = false }: { groupId: number; allowEmpty?: boolean }) =>
            apiClient.post<GroupHealthGroupView>(`/api/v1/group/health/${groupId}/attempts/exclude-failed`, { allow_empty: allowEmpty }),
        onSuccess: () => invalidateGroupHealth(queryClient),
        onError: (error) => logger.error('batch group health exclusion failed:', error),
    });
}

export function useProbeAndRestoreGroupHealthItem() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: ({ groupId, itemId }: { groupId: number; itemId: number }) =>
            apiClient.post<GroupHealthRecoveryResult>(`/api/v1/group/health/${groupId}/items/${itemId}/probe-and-restore`, {}),
        onSuccess: () => invalidateGroupHealth(queryClient),
    });
}

export function useProbeAndRestoreAllGroupHealthItems() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: ({ groupId }: { groupId: number }) =>
            apiClient.post<GroupHealthBatchRecoveryResult>(`/api/v1/group/health/${groupId}/items/probe-and-restore`, {}),
        onSuccess: () => invalidateGroupHealth(queryClient),
    });
}

export function useForceRestoreGroupHealthItem() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: ({ groupId, itemId }: { groupId: number; itemId: number }) =>
            apiClient.post<GroupHealthRecoveryResult>(`/api/v1/group/health/${groupId}/items/${itemId}/restore`, {}),
        onSuccess: () => invalidateGroupHealth(queryClient),
    });
}

export function useGroupHealthList() {
    const { enabled } = useGroupHealthEnabled();
    return useQuery({
        queryKey: ['group-health', 'list'],
        queryFn: async () => apiClient.get<GroupHealthGroupView[]>('/api/v1/group/health/list'),
        select: (data) => data.map(normalizeView),
        enabled,
        refetchInterval: (query) => {
            const data = query.state.data as GroupHealthGroupView[] | undefined;
            return data?.some((item) => item.latest?.status === 'running') ? 5000 : 30000;
        },
    });
}

export function useGroupHealth(groupId: number | null) {
    const { enabled } = useGroupHealthEnabled();
    return useQuery({
        queryKey: ['group-health', 'detail', groupId],
        queryFn: async () => apiClient.get<GroupHealthGroupView>(`/api/v1/group/health/${groupId}`),
        select: normalizeView,
        enabled: enabled && groupId != null && groupId > 0,
        refetchInterval: (query) => {
            const data = query.state.data as GroupHealthGroupView | undefined;
            return data?.latest?.status === 'running' ? 5000 : 30000;
        },
    });
}

export function useRunGroupHealth() {
    const queryClient = useQueryClient();
    const { enabled } = useGroupHealthEnabled();
    return useMutation({
        mutationFn: async ({ groupId, probeMode }: RunGroupHealthRequest) => {
            if (!enabled) throw new Error('Group health checks are disabled');
            return apiClient.post<RunGroupHealthAccepted>(
                `/api/v1/group/health/${groupId}/run`,
                probeMode ? { probe_mode: probeMode } : {},
            );
        },
        onSuccess: () => invalidateGroupHealth(queryClient),
        onError: (error) => logger.error('group health run failed:', error),
    });
}

export function useRunAllGroupHealth() {
    const queryClient = useQueryClient();
    const { enabled } = useGroupHealthEnabled();
    return useMutation({
        mutationFn: async ({ probeMode }: RunAllGroupHealthRequest = {}) => {
            if (!enabled) throw new Error('Group health checks are disabled');
            return apiClient.post<RunGroupHealthAccepted>(
                '/api/v1/group/health/run-all',
                probeMode ? { probe_mode: probeMode } : {},
            );
        },
        onSuccess: () => invalidateGroupHealth(queryClient),
        onError: (error) => logger.error('group health run-all failed:', error),
    });
}

File: internal/op/group_health.go
package op

import (
	"context"
	"errors"
	"fmt"
	"sort"
	"time"

	"github.com/bestruirui/octopus/internal/db"
	"github.com/bestruirui/octopus/internal/model"
	"gorm.io/gorm"
)

type GroupHealthRepository struct{}

func NewGroupHealthRepository() *GroupHealthRepository {
	return &GroupHealthRepository{}
}

func (r *GroupHealthRepository) CreateRunningSnapshot(ctx context.Context, group model.Group, probeMode model.GroupHealthProbeMode) (*model.GroupHealthSnapshot, error) {
	snapshot := &model.GroupHealthSnapshot{
		GroupID:      group.ID,
		GroupName:    group.Name,
		GroupMode:    group.Mode,
		ProbeMode:    probeMode,
		RequestModel: group.Name,
		Status:       model.GroupHealthStatusRunning,
		StartedAt:    time.Now(),
	}
	if err := db.GetDB().WithContext(ctx).Create(snapshot).Error; err != nil {
		return nil, err
	}
	return snapshot, nil
}

func (r *GroupHealthRepository) AppendAttempt(ctx context.Context, snapshotID int, attempt model.GroupHealthAttempt) error {
	attempt.SnapshotID = snapshotID
	return db.GetDB().WithContext(ctx).Create(&attempt).Error
}

func (r *GroupHealthRepository) FinishSnapshot(ctx context.Context, snapshotID int, status model.GroupHealthStatus, successfulChannelID *int, durationMS int64, message string, finishedAt time.Time) error {
	update := map[string]any{
		"status":      status,
		"finished_at": finishedAt,
		"duration_ms": durationMS,
		"message":     message,
	}
	if successfulChannelID != nil {
		update["successful_channel_id"] = *successfulChannelID
	} else {
		update["successful_channel_id"] = nil
	}
	return db.GetDB().WithContext(ctx).
		Model(&model.GroupHealthSnapshot{}).
		Where("id = ?", snapshotID).
		Updates(update).Error
}

func (r *GroupHealthRepository) GetLatestSnapshotByGroupID(ctx context.Context, groupID int) (*model.GroupHealthSnapshot, error) {
	var snapshot model.GroupHealthSnapshot
	err := db.GetDB().WithContext(ctx).
		Preload("Attempts", func(tx *gorm.DB) *gorm.DB {
			return tx.Order("priority ASC, id ASC")
		}).
		Where("group_id = ?", groupID).
		Order("id DESC").
		First(&snapshot).Error
	if err != nil {
		return nil, err
	}
	return &snapshot, nil
}

func (r *GroupHealthRepository) GetAttemptByID(ctx context.Context, attemptID int) (*model.GroupHealthAttempt, error) {
	var attempt model.GroupHealthAttempt
	if err := db.GetDB().WithContext(ctx).First(&attempt, attemptID).Error; err != nil {
		return nil, err
	}
	return &attempt, nil
}

func (r *GroupHealthRepository) GetSnapshotByID(ctx context.Context, snapshotID int) (*model.GroupHealthSnapshot, error) {
	var snapshot model.GroupHealthSnapshot
	if err := db.GetDB().WithContext(ctx).First(&snapshot, snapshotID).Error; err != nil {
		return nil, err
	}
	return &snapshot, nil
}

func (r *GroupHealthRepository) ListLatestSnapshots(ctx context.Context) ([]model.GroupHealthSnapshot, error) {
	var snapshots []model.GroupHealthSnapshot
	err := db.GetDB().WithContext(ctx).
		Preload("Attempts", func(tx *gorm.DB) *gorm.DB {
			return tx.Order("priority ASC, id ASC")
		}).
		Where("id IN (?)",
			db.GetDB().Model(&model.GroupHealthSnapshot{}).
				Select("MAX(id)").
				Group("group_id"),
		).
		Order("group_name ASC").
		Find(&snapshots).Error
	return snapshots, err
}

func (r *GroupHealthRepository) GetRunningSnapshotByGroupID(ctx context.Context, groupID int) (*model.GroupHealthSnapshot, error) {
	var snapshot model.GroupHealthSnapshot
	err := db.GetDB().WithContext(ctx).
		Where("group_id = ? AND status = ?", groupID, model.GroupHealthStatusRunning).
		Order("id DESC").
		First(&snapshot).Error
	if err != nil {
		return nil, err
	}
	return &snapshot, nil
}

func (r *GroupHealthRepository) ListGroupHealthViews(ctx context.Context) ([]model.GroupHealthGroupView, error) {
	var groups []model.Group
	if err := db.GetDB().WithContext(ctx).Preload("Items").Find(&groups).Error; err != nil {
		return nil, err
	}
	var err error
	snapshots, err := r.ListLatestSnapshots(ctx)
	if err != nil {
		return nil, err
	}
	latestByGroupID := make(map[int]model.GroupHealthSnapshot, len(snapshots))
	for _, snapshot := range snapshots {
		latestByGroupID[snapshot.GroupID] = snapshot
	}
	views := make([]model.GroupHealthGroupView, 0, len(groups))
	for _, group := range groups {
		view := model.GroupHealthGroupView{
			GroupID:   group.ID,
			GroupName: group.Name,
			GroupMode: group.Mode,
		}
		if snapshot, ok := latestByGroupID[group.ID]; ok {
			copySnapshot := snapshot
			view.Latest = &copySnapshot
		}
		if err := r.enrichView(ctx, &view, group.Items); err != nil {
			return nil, err
		}
		views = append(views, view)
	}
	sort.Slice(views, func(i, j int) bool {
		if views[i].GroupName != views[j].GroupName {
			return views[i].GroupName < views[j].GroupName
		}
		return views[i].GroupID < views[j].GroupID
	})
	return views, nil
}

func (r *GroupHealthRepository) GetGroupHealthViewByID(ctx context.Context, groupID int) (*model.GroupHealthGroupView, error) {
	var group model.Group
	if err := db.GetDB().WithContext(ctx).Preload("Items").First(&group, groupID).Error; err != nil {
		return nil, err
	}
	view := &model.GroupHealthGroupView{
		GroupID:   group.ID,
		GroupName: group.Name,
		GroupMode: group.Mode,
	}
	snapshot, err := r.GetLatestSnapshotByGroupID(ctx, groupID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			if enrichErr := r.enrichView(ctx, view, group.Items); enrichErr != nil {
				return nil, enrichErr
			}
			return view, nil
		}
		return nil, err
	}
	view.Latest = snapshot
	if err := r.enrichView(ctx, view, group.Items); err != nil {
		return nil, err
	}
	return view, nil
}

func (r *GroupHealthRepository) enrichView(ctx context.Context, view *model.GroupHealthGroupView, items []model.GroupItem) error {
	itemByID := make(map[int]model.GroupItem, len(items))
	for _, item := range items {
		itemByID[item.ID] = item
		if item.ExcludedAt == nil {
			view.ActiveItemCount++
		}
	}
	if view.Latest != nil {
		for i := range view.Latest.Attempts {
			item, ok := itemByID[view.Latest.Attempts[i].GroupItemID]
			switch {
			case !ok:
				view.Latest.Attempts[i].MembershipState = model.GroupHealthMembershipStateMissing
			case item.ExcludedAt != nil:
				view.Latest.Attempts[i].MembershipState = model.GroupHealthMembershipStateExcluded
			default:
				view.Latest.Attempts[i].MembershipState = model.GroupHealthMembershipStateActive
			}
		}
	}

	for _, item := range items {
		if item.ExcludedAt == nil {
			continue
		}
		var channel model.Channel
		channelName := fmt.Sprintf("channel-%d", item.ChannelID)
		channelEnabled := false
		if err := db.GetDB().WithContext(ctx).First(&channel, item.ChannelID).Error; err == nil {
			channelName, channelEnabled = channel.Name, channel.Enabled
		} else if !errors.Is(err, gorm.ErrRecordNotFound) {
			return err
		}
		excluded := model.GroupHealthExcludedItem{
			ID: item.ID, GroupID: item.GroupID, ChannelID: item.ChannelID,
			ChannelName: channelName, ChannelEnabled: channelEnabled, ModelName: item.ModelName,
			Priority: item.Priority, Weight: item.Weight, ExcludedAt: item.ExcludedAt,
			ExcludedByAttemptID: item.ExcludedByAttemptID,
		}
		var attempt model.GroupHealthAttempt
		query := db.GetDB().WithContext(ctx).Where("group_item_id = ? AND status = ?", item.ID, model.GroupHealthAttemptStatusFailed)
		if item.ExcludedByAttemptID != nil {
			query = query.Where("id = ?", *item.ExcludedByAttemptID)
		}
		if err := query.Order("id DESC").First(&attempt).Error; err == nil {
			excluded.HTTPStatus = attempt.HTTPStatus
			excluded.DurationMS = attempt.DurationMS
			excluded.ErrorMessage = attempt.ErrorMessage
		} else if !errors.Is(err, gorm.ErrRecordNotFound) {
			return err
		}
		view.ExcludedItems = append(view.ExcludedItems, excluded)
	}
	sort.Slice(view.ExcludedItems, func(i, j int) bool { return view.ExcludedItems[i].ID < view.ExcludedItems[j].ID })
	return nil
}
